File: frontend/src/services/api/ocr.ts
import { api } from "./client";
import type { ApiResponse } from "@/types";

export interface OCRResult {
  id: string;
  filename: string;
  status: string;
  document_type: string;
  raw_text: string;
  average_confidence: number;
  provider?: string | null;
  boxes: Array<{
    text: string;
    confidence: number;
    bbox: number[][];
    page?: number;
  }>;
  tables: unknown[];
  key_values: Record<string, string>;
  layout: Record<string, unknown>;
  structured_json: Record<string, unknown>;
  metrics: Record<string, unknown>;
  linked_document_id?: string | null;
  thumbnail_path?: string | null;
  created_at?: string | null;
}

export interface VisionObject {
  id?: string;
  label: string;
  confidence: number;
  bbox?: number[] | null;
  model_name: string;
}

export interface VisionAnalysis {
  id: string;
  filename: string;
  status: string;
  caption?: string | null;
  scene_description?: string | null;
  chart_summary?: string | null;
  screenshot_explanation?: string | null;
  provider?: string | null;
  metrics: Record<string, unknown>;
  objects: VisionObject[];
  created_at?: string | null;
}

export const ocrApi = {
  upload(file: File) {
    const form = new FormData();
    form.append("file", file);
    return api.post<ApiResponse<OCRResult>>("/ocr/upload", form, {
      headers: { "Content-Type": "multipart/form-data" },
    });
  },
  get(id: string) {
    return api.get<ApiResponse<OCRResult>>(`/ocr/${id}`);
  },
  list(params?: { q?: string; document_type?: string; limit?: number }) {
    return api.get<
      ApiResponse<{ items: OCRResult[]; total: number; limit: number; offset: number }>
    >("/ocr", { params });
  },
  search(q: string) {
    return api.get<
      ApiResponse<{ items: OCRResult[]; total: number; limit: number; offset: number }>
    >("/ocr/search", { params: { q } });
  },
};

export const visionApi = {
  analyze(file: File) {
    const form = new FormData();
    form.append("file", file);
    return api.post<ApiResponse<VisionAnalysis>>("/vision/analyze", form, {
      headers: { "Content-Type": "multipart/form-data" },
    });
  },
  detect(file: File) {
    const form = new FormData();
    form.append("file", file);
    return api.post<ApiResponse<VisionAnalysis>>("/vision/detect", form, {
      headers: { "Content-Type": "multipart/form-data" },
    });
  },
  history(params?: { limit?: number; offset?: number }) {
    return api.get<
      ApiResponse<{
        items: Array<{
          id: string;
          filename: string;
          caption?: string;
          object_count: number;
          created_at?: string;
        }>;
        total: number;
      }>
    >("/vision/history", { params });
  },
  get(id: string) {
    return api.get<ApiResponse<VisionAnalysis>>(`/vision/${id}`);
  },
};

File: backend/app/schemas/ocr.py
"""OCR & Vision API schemas."""

from __future__ import annotations

from typing import Any, Optional

from pydantic import BaseModel, ConfigDict, Field


class OCRResultOut(BaseModel):
    id: str
    filename: str
    status: str
    document_type: str
    raw_text: str = ""
    average_confidence: float = 0.0
    provider: Optional[str] = None
    boxes: list[Any] = Field(default_factory=list)
    tables: list[Any] = Field(default_factory=list)
    key_values: dict[str, Any] = Field(default_factory=dict)
    layout: dict[str, Any] = Field(default_factory=dict)
    structured_json: dict[str, Any] = Field(default_factory=dict)
    metrics: dict[str, Any] = Field(default_factory=dict)
    linked_document_id: Optional[str] = None
    thumbnail_path: Optional[str] = None
    created_at: Optional[str] = None


class OCRListOut(BaseModel):
    items: list[dict[str, Any]]
    total: int
    limit: int
    offset: int


class VisionObjectOut(BaseModel):
    model_config = ConfigDict(protected_namespaces=())

    id: Optional[str] = None
    label: str
    confidence: float
    bbox: Optional[list[Any]] = None
    model_name: str = "yolo"


class VisionAnalysisOut(BaseModel):
    id: str
    filename: str
    status: str
    caption: Optional[str] = None
    scene_description: Optional[str] = None
    chart_summary: Optional[str] = None
    screenshot_explanation: Optional[str] = None
    provider: Optional[str] = None
    metrics: dict[str, Any] = Field(default_factory=dict)
    objects: list[VisionObjectOut] = Field(default_factory=list)
    created_at: Optional[str] = None


class VisionHistoryOut(BaseModel):
    items: list[dict[str, Any]]
    total: int
    limit: int
    offset: int
